extensions/functions/src/main/java/io/crate/window/OffsetValueFunctions.java

```diff
@@ -45,34 +46,173 @@
+                expression.setNextRow(row);
+            }
+            return args[0].value();
+        }
+    }
+
+    private final Supplier<OffsetDirection> FORWARD = () -> new OffsetDirection() {
+
+        @Override
+        int getTargetIndex(int idxInPartition, int offset) {
+            return idxInPartition + offset;
+        }
+
+        @Override
+        Object getValueAtOffsetIgnoringNulls(int idxInPartition,
+                                             int offset,
+                                             WindowFrameState currentFrame,
+                                             List<? extends CollectExpression<Row, ?>> expressions,
+                                             Input[] args) {
+            if (cachedNonNullIndex == null) {
+                cachedNonNullIndex = findNonNullOffsetFromCurrentIndex(idxInPartition, offset, currentFrame, expressions, args);
+            } else {
+                var curValue = getValueAtTargetIndex(idxInPartition, currentFrame, expressions, args);
+                if (curValue != null) {
+                    moveCacheToNextNonNull(currentFrame, expressions, args);
+                }
+            }
+            return getValueAtTargetIndex(cachedNonNullIndex, currentFrame, expressions, args);
+        }
+    };
+
+    private final Supplier<OffsetDirection> BACKWARD = () -> new OffsetDirection() {
+
+        @Override
+        int getTargetIndex(int idxInPartition, int offset) {
+            return idxInPartition - offset;
+        }
+
+        @Override
+        Object getValueAtOffsetIgnoringNulls(int idxInPartition,
+                                             int offset,
+                                             WindowFrameState currentFrame,
+                                             List<? extends CollectExpression<Row, ?>> expressions,
+                                             Input[] args) {
+            if (cachedNonNullIndex == null) {
+                cachedNonNullIndex = findNonNullOffsetFromCurrentIndex(idxInPartition, offset, currentFrame, expressions, args);
+            } else {
+                var prevValue = getValueAtTargetIndex(idxInPartition - 1, currentFrame, expressions, args);
+                if (prevValue != null) {
+                    moveCacheToNextNonNull(currentFrame, expressions, args);
+                }
+            }
+            return getValueAtTargetIndex(cachedNonNullIndex, currentFrame, expressions, args);
+        }
+    };
+
+    private OffsetDirection resolveOffsetDirection(int offset) {
+        if (LAG_NAME.equals(signature.getName().name())) {
+            if (offset >= 0) {
+                return BACKWARD.get();
+            } else {
+                return FORWARD.get();
+            }
+        } else {
+            if (offset >= 0) {
+                return FORWARD.get();
+            } else {
+                return BACKWARD.get();
+            }
+        }
```

Comment:
Might be possible to reduce branching at runtime by parameterizing the constructor with a multpilier that is either `1` or `-1`. To do something like `return ((offset * directionMultiplier) > 0)` here.